Add framework for implementation of support for AVX512VNNI

stweil opened the pull request:
Using AVX512VNNI instructions is still missing.

stweil committed:
Implement intSimdMatrixAVX512VNNI (dummy)

This dummy implementation just copied the code from AVX2.

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Add unittest for AVX512VNNI

Signed-off-by: Stefan Weil <sw@weilnetz.de>

stweil committed:
Test whether OS supports AVX512

Signed-off-by: Stefan Weil <sw@weilnetz.de>

amitdo committed:
Add the needed changes to support AVX512VNNI

stweil commented on Makefile.am:
```diff
@@ -170,6 +170,14 @@ libtesseract_la_LIBADD += libtesseract_avx512.la
 endif
 
+if HAVE_AVX512VNNI
+libtesseract_avx512vnni_la_CXXFLAGS = -mavx512vnni -mavx512vl
```
I am curious: what is the use of `-mavx512vl`? Should it be checked in `configure`, too?

amitdo replied:
It does not compile without the second flag. It is needed becaIuse the code uses ymm registers (256 bit) instead of zmm registers (512 bit).

Yes, we can also add another check in configure.

Note that AFAIK all intel's CPUs that support AVX512VNNI also support AVXVL.

stweil replied:
Thanks for explaining. Maybe we don't need the additional check then.

amitdo commented:
Ready for testing.

I suggest to backup the home directory and any other important stuff, in case this patch will trigger a hard crash (reboot).

stweil commented:
> I suggest to backup the home directory and any other important stuff, in case this patch will trigger a hard crash (reboot).

Will it? I cannot test it before August 21 because I currently don't have access to the required hardware.

amitdo commented:
>Will it?

I don't think so...

amitdo commented:
Stefan, will you able to test the code today or in in the upcoming days?

stweil commented:
> Stefan, will you able to test the code today or in in the upcoming days?

I just finished a first run of `intsimdmatrix_test` with the new code. It works, but I was surprised because the performance is identical to the AVX2 implementation.

stweil commented:
Support for sw build is still missing. Therefore CI currently fails.

amitdo commented:
>I was surprised because the performance is identical to the AVX2 implementation.

This is disappointing.

We can also remove the 'ones' vector. The vnni version does not need it.

amitdo commented:
Did you fetched and merged my changes locally?

```
git switch avx512-vnni
git log -1
```

Do you see my commit?

stweil commented:
> We can also remove the 'ones' vector. The vnni version does not need it.

I removed it locally. But that did not change the generated code. Obviously the compiler also removed it.

amitdo commented:
@stweil,

Do you have access to anther machine (non-notebook) that supports AVX512VNNI ?

stweil commented:
No, I have only sometimes access to a notebook which supports it.
